Lib/typing.py

```diff
@@ -1290,13 +1290,30 @@ def __call__(self, *args, **kwargs):
         if not self._inst:
             raise TypeError(f"Type {self._name} cannot be instantiated; "
                             f"use {self.__origin__.__name__}() instead")
-        result = self.__origin__(*args, **kwargs)
+        clas = self.__origin__
+        is_python_type = clas.__new__ is type.__new__
```

Comment:
This doesn't seem to do what you want it to do:
```
>>> class A: pass
... 
>>> A.__new__ is type.__new__
False
```

This PR doesn't include tests; to get it merged we'll definitely need tests.

I'm also not convinced we should change this at all, since it feels like a risky change at this point.

Reply:
You may want to compare against `object.__new__` instead, but that still doesn't directly check for "Python types", and in any case I'm not sure it's a good idea to branch on whether it's a Python type.

```
>>> from typing import TypeVar
>>> TypeVar.__new__
<built-in method __new__ of type object at 0x14c817e30>
>>> A.__new__ is object.__new__
True
```

Reply:
Yeah, Alex had the same concern, I just pushed a better fix instead of `type.__new__`. 

> I'm also not convinced we should change this at all, since it feels like a risky change at this point.

Very much agreeing there, but there does seem to be some demand from the typing users. If this causes a regression somewhere, we could always roll back the changes!

Reply:
> If this causes a regression somewhere, we could always roll back the changes!

If we release this in 3.14.0, then backward compatibility will again mean we're stuck with the change. CPython is deep enough in the stack that we can't afford to just go back and forth.

Reply:
> You may want to compare against `object.__new__` instead, but that still doesn't directly check for "Python types", and in any case I'm not sure it's a good idea to branch on whether it's a Python type.
> 
> ```
> >>> from typing import TypeVar
> >>> TypeVar.__new__
> <built-in method __new__ of type object at 0x14c817e30>
> >>> A.__new__ is object.__new__
> True
> ```

Oh yeah, that sounds like the right approach, my bad!

> If we release this in 3.14.0, then backward compatibility will again mean we're stuck with the change. CPython is deep enough in the stack that we can't afford to just go back and forth.

Hmm, maybe we should bring this to Discourse.

Reply:
> Very much agreeing there, but there does seem to be some demand from the typing users.

Just because there's a problem doesn't mean that this is the best fix. `__orig_class__` is completely undocumented and the last time we discussed potentially documenting it we decided not to because it sometimes has a counterintuitive value. For other typing-specific dunders our approach has generally been to expose public introspection helpers so that we're easily able to refactor internals later without breaking any public interface (e.g. `get_protocol_members()`, `types.get_original_bases`, etc.).